core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+                }),
+        }
+    }
+
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
+    fn perform_handshake(
+        &self,
+        mut stream: TcpStream,
+    ) -> Result<(TcpStream, Cryptographer, PublicKey), Error> {
+        let crypto = self.exchange_greetings(&mut stream)?;
+        trace!("Sent public key to other peer.");
+        let mut size_buf = 0_u8;
+        stream.read_exact(std::slice::from_mut(&mut size_buf))?;
+        if (size_buf as usize) >= MAX_HANDSHAKE_LENGTH {
+            return Err(Error::HandshakeLength(size_buf as usize));
+        }
+        let mut null_bytes = vec![0_u8; size_buf as usize];
+        stream.read_exact(&mut null_bytes)?;
+        let other_public_key = Decode::decode(&mut crypto.decrypt(null_bytes)?.as_slice())?;
+        trace!("Completed handshake with {}.", other_public_key);
+        Ok((stream, crypto, other_public_key))
+    }
+
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
+    fn exchange_greetings(&self, stream: &mut TcpStream) -> Result<Cryptographer, Error> {
+        Garbage::generate().write(stream)?;
+        let mut crypto = Cryptographer::default().map_err(CryptographicError::Ursa)?;
+        stream.write_all(crypto.public_key.0.as_slice())?;
+        trace!("Sent hello.");
+        Garbage::read(stream)?;
+        let mut key = [0_u8; 32];
+        stream.read_exact(&mut key)?;
+        crypto.derive_shared_key(&UrsaPublicKey(Vec::from(key)))?;
+        trace!("Received hello.");
+        let data = crypto.encrypt(self.public_key.encode())?;
+        let mut data_buf = Vec::<u8>::with_capacity(data.len() + 1);
+        data_buf.push(u8::try_from(data.len()).expect("data length doesn't fit into byte."));
+        data_buf.extend_from_slice(data.as_slice());
+        stream.write_all(&data_buf)?;
+        Ok(crypto)
+    }
+}
+
+#[allow(clippy::expect_used)]
+fn finish_connection(
+    crypto: Cryptographer,
+    mut stream: TcpStream,
+    connected_to_peers: &mut HashMap<PublicKey, PeerConnection>,
+    other_public_key: &PublicKey,
+) -> eyre::Result<()> {
+    let encrypted = crypto
+        .encrypt(NetworkMessage::ConnectionCheck(42).encode())
+        .expect("We should always be able to encrypt.");
+    stream.write_all(
+        &u32::try_from(encrypted.len())
+            .expect("decoded length doesn't fit into `u32`")
+            .to_le_bytes(),
+    )?;
+    stream.write_all(&encrypted)?;
+    let mut packet_size_buf = [0_u8; 4];
+    stream.read_exact(&mut packet_size_buf)?;
+    let packet_size = u32::from_le_bytes(packet_size_buf);
+    let mut buf = vec![0_u8; packet_size as usize];
+    stream.read_exact(&mut buf)?;
+    let data = crypto.decrypt(buf)?;
+    if let ConnectionCheck(_) = Decode::decode(&mut data.as_slice())? {}
+    else {
+        error!("Actually not established connection to peer {other_public_key}."); // Rust errors are dumb, TODO invent an error type for this case.
+    }
```

Comment:
Actually, that's precisely why the Rust Errors are the way they are: you're supposed to produce new types for new kinds of events. And this would let you catch these problems with a neat syntax.